core/src/main/scala/kafka/network/SocketServer.scala

```diff
@@ -538,14 +542,20 @@ private[kafka] class Acceptor(val endPoint: EndPoint,
                               val recvBufferSize: Int,
                               brokerId: Int,
                               connectionQuotas: ConnectionQuotas,
-                              metricPrefix: String) extends AbstractServerThread(connectionQuotas) with KafkaMetricsGroup {
+                              metricPrefix: String,
+                              time: Time) extends AbstractServerThread(connectionQuotas) with KafkaMetricsGroup {
 
   private val nioSelector = NSelector.open()
   val serverChannel = openServerSocket(endPoint.host, endPoint.port)
   private val processors = new ArrayBuffer[Processor]()
   private val processorsStarted = new AtomicBoolean
   private val blockedPercentMeter = newMeter(s"${metricPrefix}AcceptorBlockedPercent",
     "blocked time", TimeUnit.NANOSECONDS, Map(ListenerMetricTag -> endPoint.listenerName.value))
+  private var currentProcessorIndex = 0
+  private[network] case class DelayedCloseSocket(socket: SocketChannel, endThrottleTimeMs: Long) extends Ordered[DelayedCloseSocket] {
+    override def compare(that: DelayedCloseSocket): Int = endThrottleTimeMs compare that.endThrottleTimeMs
+  }
+  private[network] val throttledSockets = new mutable.PriorityQueue[DelayedCloseSocket]()
```

Comment:
my comment on this may have gotten lost in the weeds, for reference:

I did not use the java DelayQueue similar to the implementation in ClientQuotaManager because this throttling implementation does not use timeout-based polling or require a synchronized data structure, and there's a bit more boilerplate needed for a DelayQueue to support timeout-based polling.

Reply:
Thanks for the clarification. I have missed that previous comment.